Lib/test/test_capi/test_abstract.py

```diff
@@ -1007,6 +1007,46 @@ def test_object_generichash(self):
         for obj in object(), 1, 'string', []:
             self.assertEqual(generichash(obj), object.__hash__(obj))
 
+    def run_iter_api_test(self, next_func):
+        for data in (), [], (1, 2, 3), [1 , 2, 3], "123":
+            with self.subTest(data=data):
+                items = []
+                it = iter(data)
+                while (item := next_func(it)) is not None:
+                    items.append(item)
+                self.assertEqual(items, list(data))
+
+        class Broken:
+            def __init__(self):
+                self.count = 0
+
+            def __next__(self):
+                if self.count < 3:
+                    self.count += 1
+                    return self.count
+                else:
+                    raise TypeError('bad type')
+
+        it = Broken()
+        self.assertEqual(next_func(it), 1)
+        self.assertEqual(next_func(it), 2)
+        self.assertEqual(next_func(it), 3)
+        with self.assertRaisesRegex(TypeError, 'bad type'):
+            next_func(it)
+
+    def test_iter_next(self):
+        from _testcapi import PyIter_Next
+        self.run_iter_api_test(PyIter_Next)
+        # CRASHES PyIter_Next(10)
+
+    def test_iter_nextitem(self):
+        from _testcapi import PyIter_NextItem
+        self.run_iter_api_test(PyIter_NextItem)
+
+        regex = "expected.*iterator.*got.*'int'"
```

Comment:
```suggestion
        regex = re.escape("expected an iterator, got 'int'")
```

Much clearer rather than those wildcards IMO. It also guarantees that in the future, we don't change error messages without checking them beforehand.

Reply:
There's a balance of checking exact error messages and leaving room for small adjustments in the error text. You'll find both kinds in the stdlib, but I think we generally prefer the latter, though I don't think it is written down anywhere.

Reply:
Yes, what I found was:

- No exception message at all (just the type),
- A too-wide regex which might not always match the exception we thought of,
- A very strict message (in general, when I add tests, I like having a robust check because sometimes, changing messages need to be carefully done just in case we might break doctest outputs (I sometimes have those kind of surprises when a message changes)).

Let's leave it like this I think.